pkg/inspector/inspector.go

```diff
@@ -87,81 +79,84 @@ func New(logger log.CtxLogger, bufferSize int) *Inspector {
 	}
 }
 
+// NewSession creates a new session in given inspector.
+// componentID is the ID of the component being inspected (connector or processor).
+func (i *Inspector) NewSession(ctx context.Context, componentID string) *Session {
+	s := &Session{
+		C:           make(chan record.Record, i.bufferSize),
+		id:          uuid.NewString(),
+		componentID: componentID,
+		logger:      i.logger.WithComponent("inspector.Session"),
+	}
+
+	i.add(s)
+	go func() {
+		<-ctx.Done()
+		i.remove(s.id)
+	}()
+
+	return s
+}
+
 // Send the given record to all registered sessions.
 // The method does not wait for consumers to get the records.
 func (i *Inspector) Send(ctx context.Context, r record.Record) {
-	// copy metadata, to prevent issues when concurrently accessing the metadata
-	var meta record.Metadata
-	if len(r.Metadata) != 0 {
-		meta = make(record.Metadata, len(r.Metadata))
-		for k, v := range r.Metadata {
-			meta[k] = v
-		}
+	// shortcut - we don't expect any sessions, so we check the atomic variable
+	// before acquiring an actual lock
+	if !i.hasSessions.Load() {
+		return
 	}
 
-	// todo optimize this, as we have locks for every record.
+	// clone record only once, the listeners aren't expected to manipulate the records
+	rClone := r.Clone()
+
 	// locks are needed to make sure the `sessions` slice
 	// is not modified as we're iterating over it
 	i.lock.Lock()
 	defer i.lock.Unlock()
 	for _, s := range i.sessions {
-		s.send(ctx, record.Record{
-			Position:  r.Position,
-			Operation: r.Operation,
-			Metadata:  meta,
-			Key:       r.Key,
-			Payload:   r.Payload,
-		})
+		s.send(ctx, rClone)
 	}
 }
 
-// NewSession creates a new session in given inspector.
-// componentID is the ID of the component being inspected (connector or processor).
-func (i *Inspector) NewSession(ctx context.Context, componentID string) *Session {
-	id := uuid.NewString()
-	s := &Session{
-		C:      make(chan record.Record, i.bufferSize),
-		id:     id,
-		logger: i.logger.WithComponent("inspector.Session"),
-		onClose: func() {
-			i.remove(id)
-			measure.InspectorsGauge.WithValues(componentID).Dec()
-		},
-		once: &sync.Once{},
+func (i *Inspector) Close() {
+	for k := range i.sessions {
+		i.remove(k)
 	}
-	measure.InspectorsGauge.WithValues(componentID).Inc()
-
-	go func() {
-		<-ctx.Done()
-		s.logger.
-			Debug(context.Background()).
-			Msgf("context done: %v", ctx.Err())
-		s.close()
-	}()
+}
 
+// add a session with given ID to this Inspector.
+func (i *Inspector) add(s *Session) {
```

Comment:
_[nit]_ Do we need to send the whole struct (and have a pointer of a pointer) or could be just `id string` (or `sessionID string`).

Reply:
Well, we need the whole struct as we're storing it, and we need to know the whole thing if we want to later remove it. `Session` could actually be passed as a value instead of a pointer, but after changing it I see the same heap allocations 😕 So I'm not sure if it's worth changing it, especially since the change would cascade down into other layers all the way to the orchestrator (e.g. [here](https://github.com/ConduitIO/conduit/blob/b7ab8484e03612fa45ff41dbc0e7349a23aa8204/pkg/orchestrator/processors.go#L109)).

Reply:
Btw, to clarify, the session creation is not in the hot path. The actual optimization in this PR is coming from the behavior of `Inspector.Send` where we have a shortcut to skip the mutex and record cloning if there are no sessions. That is in the hot path and actually makes a difference.